File: backend/services/post_service.js
import { query, notFound } from "../utils/errors.js";
import { getJourney } from "./journey_service.js";

// Public snapshots are written by a checked SQL transaction, not by copying
// arbitrary request JSON. Completion, ownership and visited stops are checked there.
export function publishJourney(db, journeyId, body) {
  return query(
    db.rpc("publish_journey", {
      journey_id_input: journeyId,
      country_input: body.country,
      caption_input: body.caption,
    }),
  );
}
export async function listPosts(db, offset, country) {
  let selection = db
    .from("published_trips")
    .select("*,published_stops(*)")
    .order("published_at", { ascending: false })
    .order("id")
    .range(offset, offset + 19);
  if (country) selection = selection.eq("country", country);
  return { items: await query(selection), offset, limit: 20 };
}
export async function getPost(db, postId) {
  const post = await query(
    db
      .from("published_trips")
      .select("*,published_stops(*)")
      .eq("id", postId)
      .maybeSingle(),
  );
  if (!post) throw notFound("Post not found.");
  return post;
}
export async function reusePost(db, userId, postId, startDate) {
  // SQL generates new IDs and resets bookings/journal details in one transaction.
  const journeyId = await query(
    db.rpc("reuse_itinerary", {
      post_id_input: postId,
      start_date_input: startDate,
    }),
  );
  return getJourney(db, userId, journeyId);
}
export async function deletePost(db, userId, postId) {
  const rows = await query(
    db
      .from("published_trips")
      .delete()
      .eq("id", postId)
      .eq("user_id", userId)
      .select("id"),
  );
  if (!rows.length) throw notFound("Post not found.");
}

File: backend/services/post_service.py
"""Public itinerary snapshots and reuse; database functions enforce publication rules."""

from fastapi import HTTPException
from utils.errors import execute
from services.journey_service import get_journey


def publish_journey(db, journey_id, body):
    return execute(
        db.rpc(
            "publish_journey",
            {
                "journey_id_input": journey_id,
                "country_input": body.country,
                "caption_input": body.caption,
            },
        )
    )


def list_posts(db, offset, country):
    query = (
        db.table("published_trips")
        .select("*,published_stops(*)")
        .order("published_at", desc=True)
        .order("id")
        .range(offset, offset + 19)
    )
    if country:
        query = query.eq("country", country)
    return {"items": execute(query), "offset": offset, "limit": 20}


def get_post(db, post_id):
    rows = execute(
        db.table("published_trips")
        .select("*,published_stops(*)")
        .eq("id", post_id)
        .limit(1)
    )
    if not rows:
        raise HTTPException(404, "Post not found.")
    return rows[0]


def delete_post(db, user_id, post_id):
    rows = execute(
        db.table("published_trips").delete().eq("id", post_id).eq("user_id", user_id)
    )
    if not rows:
        raise HTTPException(404, "Post not found.")


def reuse_post(db, user_id, post_id, start_date):
    # New IDs, reset bookings, and blank journals are created atomically in SQL.
    journey_id = execute(
        db.rpc(
            "reuse_itinerary",
            {"post_id_input": post_id, "start_date_input": start_date},
        )
    )
    return get_journey(db, user_id, journey_id)
